process_manager/tables.py

```diff
@@ -43,6 +54,28 @@ class Meta:  # noqa: D106
     select = tables.CheckBoxColumn(
         accessor="uuid",
         verbose_name="Select",
-        checked="checked",
-        attrs={"th__input": {"onclick": "toggle(this)"}},
+        attrs={
+            "th__input": {
+                "id": "header-checkbox",
+                "hx-preserve": "true",
+                "_": header_checkbox_hyperscript,
+            }
+        },
     )
+
+    def render_select(self, value: str) -> str:
+        """Customise behaviour of checkboxes in the select column.
+
+        Overriding the default render method for this column is required as the
+        hx-preserve attitribute requires all elements to have unique id values. We also
+        need to add the hyperscript required for the header checkbox behaviour.
+
+        Called during table rendering.
+
+        Args:
+          value: uuid from the table row data
+        """
+        return mark_safe(
```

Comment:
So `render_select` is magically called when rendering `self.select = tables.CheckBoxCol...`, and creates a custom `id` for the row checkbox, so that `hx-preserve` can preserve each row cbox on refresh.

What is `mark_safe` doing here exactly?

Reply:
https://docs.djangoproject.com/en/5.1/ref/utils/#django.utils.safestring.mark_safe It tells the Django templating system that the html generated is safe and isn't vurnerable to injection attacks so it can just be rendered directly. Otherwise the templating system by default will try to escape everything.